### PR TITLE
[DO NOT MERGE] Preview functionality

### PR DESCRIPTION
**Pull request by 3ch023**

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            return pageProperties.getInherited(DemandwareCommerceConstants.PN_DWRE_INSTANCE_ID, previewInstanceId);
+        }
+        return previewInstanceId;
+    }
+    
+    private Page getPage(SlingHttpServletRequest request) {
+        String path = getRequestedPath(request);
+        
+        ResourceResolver resourceResolver = request.getResourceResolver();
+        Resource resource = resourceResolver.resolve(path);
```
Comment (3ch023) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java]: I suppose we don't need the Resource here

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.felix.scr.annotations.Component;
+import org.apache.felix.scr.annotations.Service;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+
+
+@Component(label = "Demandware Component Instance Id Service")
+@Service()
+public class DemandwareInstanceIdImpl implements DemandwareInstanceId {
```
Comment (3ch023) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java]: the name should indicate what this service does, I would rename it to something like "DemandwareEndpointProvider". Currently, it is confusing that DemandwareInstanceId provides endpoints.

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    
+    @Override
+    public String getEndPoint(DemandwareClientProvider clientProvider, Page page) {
```
Comment (3ch023) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java]: DemandwareClientProvider can be injected in this service as a @Reference, then you don't need to pass it as a parameter

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java (hunk 1 above)
Comment (allwardt) [reply bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java]: ja stimmt - habe es entfernt

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java (hunk 3 above)
Comment (3ch023) [reply bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java]: I checked code once more and noticed we now fetch DemandwareClient twice in PreviewServiceImpl.deliver method:
First time we have it in PreviewServiceImpl, line 170:
`final HttpClientBuilder httpClientBuilder = clientProvider.getDefaultClient().getHttpClientBuilder();`
and the same happens in DemandwareInstanceIdImpl.getEndPoint

Can you change .getDefaultClient() to fetching concrete client? 
`final String instanceId = getInstanceId(containingPage);`    
`final DemandwareClient dwClient = clientProvider.getDemandwareClientByInstanceId(instanceId);`
`final HttpClientBuilder httpClientBuilder = dwClient.getHttpClientBuilder();`

then you can reuse the dwClient in line 186:
`requestBuilder.setUri(DemandwareClient.DEFAULT_SCHEMA + dwClient.getEndpoint() + previewEndpoint);`

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java [hunk 4]
```diff
@@ -74,44 +75,49 @@
 public abstract class AbstractOCAPITransportPlugin extends AbstractTransportHandlerPlugin {
     static final String ACCESS_TOKEN_PROPERTY = "accessTokenProvider";
     static final String BEARER_AUTHENTICATION_FORMAT = "Bearer %s";
-
+
```
Comment (3ch023) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java:78]: Please remove these spaces

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/StaticContentProxyServlet.java [hunk 5]
```diff
@@ -57,13 +53,16 @@
...
     protected void doGet(SlingHttpServletRequest request, SlingHttpServletResponse response) throws ServletException,
             IOException {
         
         RequestPathInfo pathInfo = request.getRequestPathInfo();
         LOG.debug("Proxy static content for {}", pathInfo.toString());
-        final String remoteUri = DemandwareClient.DEFAULT_SCHEMA + clientProvider.getDemandwareClientByInstanceId(clientProvider.getInstanceId(getPage(request))) + pathInfo.getResourcePath() +
+        final String remoteUri = DemandwareClient.DEFAULT_SCHEMA + demandwareInstanceId.getEndpoint(clientProvider, request) + pathInfo.getResourcePath() +
                 "." + pathInfo.getExtension() + pathInfo.getSuffix();
         
         final CloseableHttpClient httpClient = clientProvider.getDefaultClient().getHttpClient();
```
Comment (3ch023) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/StaticContentProxyServlet.java]: Same as in PreviewService we fetch DemandwareClient here twice

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/PreviewServiceImpl.java [hunk 6]
```diff
@@ -19,6 +19,7 @@
 import com.adobe.cq.commerce.demandware.DemandwareClient;
```
Comment (3ch023) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/PreviewServiceImpl.java:19]: credentialsProvider is initialized in activate method, which means it doesn't support 2 SFCC instances

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        String previewInstanceId = getInstanceId(page);
+        return previewInstanceId != null ? clientProvider.getDemandwareClientByInstanceId(previewInstanceId).getEndpoint() : clientProvider.getDefaultClient().getEndpoint();
```
Comment (3ch023) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/DemandwareInstanceIdImpl.java]: We should not fallback to the default client here, if I am fetching content from SFCC in US and can't find US store I still should not try to fetch it from EU

**Pull request merged**